kernel/src/mm/physical_memory_manager.rs

```diff
@@ -160,33 +221,50 @@ impl PhysicalMemoryManager {
     /// space large enough to hold all our metadata.
     /// - Lastly store our metadata there, and mark pages as unused or kernel.
     pub fn from_device_tree(device_tree: &DeviceTree, page_size: usize) -> Self {
-        let page_count = Self::count_pages(device_tree, page_size);
+        let available_regions = Self::available_memory_regions::<10>(device_tree);
+        
+        let page_count = Self::count_pages(&available_regions, page_size);
         let metadata_size = page_count * mem::size_of::<PhysicalPage>();
         let pages_needed = Self::align_up(metadata_size, page_size) / page_size;
 
-        let metadata_addr =
-            Self::find_contiguous_unused_pages(device_tree, pages_needed, page_size).unwrap();
+        let metadata_addr = Self::find_large_region(&available_regions, metadata_size).unwrap();
 
         let metadata: &mut [PhysicalPage] =
             unsafe { core::slice::from_raw_parts_mut(metadata_addr as *mut _, page_count) };
 
-        device_tree.for_all_memory_regions(|regions| {
-            let physical_pages = regions
-                .flat_map(|(start, size)| (start..start + size).step_by(page_size))
-                .filter(|addr| !device_tree.is_used(*addr))
-                .map(|base| {
-                    Self::phys_addr_to_physical_page(
-                        base,
-                        metadata_addr,
-                        metadata_addr + pages_needed * page_size,
-                        device_tree,
-                    )
-                });
 
-            for (i, page) in physical_pages.enumerate() {
-                metadata[i] = page;
-            }
-        });
+        //device_tree.for_all_memory_regions(|regions| {
+        //    let physical_pages = regions
+        //        .flat_map(|(start, size)| (start..start + size).step_by(page_size))
+        //        .filter(|addr| !device_tree.is_used(*addr))
+        //        .map(|base| {
+        //            Self::phys_addr_to_physical_page(
+        //                base,
+        //                metadata_addr,
+        //                metadata_addr + pages_needed * page_size,
+        //                device_tree,
+        //            )
+        //        });
+
+        //    for (i, page) in physical_pages.enumerate() {
+        //        metadata[i] = page;
+        //    }
+        //});
+        let physical_pages = available_regions
+            .iter()
+            .filter_map(|maybe_region| maybe_region.map(|region| region))
+            .flat_map(|region| (region.start..region.end).step_by(page_size))
+            .map(|base| {
+                Self::phys_addr_to_physical_page(
```

Comment:
Be careful, now we no longer use the PageKind::Reserved variant, we should remove it (since we only put metadata / allocatable memory in the page allocator now)